pgmax/groups/logical.py

```diff
@@ -0,0 +1,169 @@
+
+        Args:
+            vars_to_starts: A dictionary that maps variables to their global starting indices
+                For an n-state variable, a global start index of m means the global indices
+                of its n variable states are m, m + 1, ..., m + n - 1
+
+        Returns:
+             LogicalWiring for the LogicalFactorGroup
+        """
+        # TODO: run in parallel?
+        import time
+
+        start = time.time()
+        relevant_state = (-self.edge_states_offset + 1) // 2
+
+        var_states_for_edges = []
+        for variable_and_num_states in self.variables_and_num_states:
+            variable, num_states = variable_and_num_states
+            this_var_states_for_edges = np.arange(
+                vars_to_starts[variable], vars_to_starts[variable] + num_states
+            )
+            var_states_for_edges.append(this_var_states_for_edges)
+        print(time.time() - start)
+
+        edges_num_states_cumsum = 0
+        parents_edge_states = []
+        children_edge_states = []
+        for factor_idx, variable_names_for_factor in enumerate(
+            self.variable_names_for_factors
+        ):
+            num_parents = len(variable_names_for_factor) - 1
+
+            # Note: edges_num_states_cumsum corresponds to the factor_to_msgs_start for the LogicalFactor
+            this_parents_edge_states = np.vstack(
+                [
+                    np.full(num_parents, fill_value=factor_idx, dtype=int),
+                    np.arange(
+                        edges_num_states_cumsum + relevant_state,
+                        edges_num_states_cumsum + 2 * num_parents,
+                        2,
+                        dtype=int,
+                    ),
+                ],
+            ).T
+            this_child_edge_state = (
+                edges_num_states_cumsum + 2 * num_parents + relevant_state
+            )
+
+            parents_edge_states.append(this_parents_edge_states)
+            children_edge_states.append(this_child_edge_state)
+            edges_num_states_cumsum += 2 * (num_parents + 1)
+        print(time.time() - start)
+
+        return logical.LogicalWiring(
+            edges_num_states=self.factor_edges_num_states,
+            var_states_for_edges=np.concatenate(var_states_for_edges),
+            parents_edge_states=np.concatenate(parents_edge_states),
+            children_edge_states=np.array(children_edge_states),
+            edge_states_offset=self.edge_states_offset,
+        )
+
+
+@dataclass(frozen=True, eq=False)
+class ORFactorGroup(LogicalFactorGroup):
+    """Class to represent a group of ORFactors.
+
+    Args:
+        edge_states_offset: Offset to go from a variable's relevant state to its other state
+            For ORFactors the edge_states_offset is 1.
+    """
+
+    edge_states_offset: int = field(init=False, default=1)
+    factor_type: Type = field(init=False, default=logical.ORFactor)
+
+    def _get_variables_to_factors(
```

Comment:
Since `factor_type` is an attribute, it looks like this function can be implemented just one within the `LogicalFactorGroup`, with factor creations done by `self.factor_type(...)`?